cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -236,6 +247,138 @@ func canInstallAndRunAtOnce(mst *initramfsMountsState) (bool, error) {
 	return kernelHasFdeSetup, nil
 }
 
+func getSnap(sysSnaps map[snap.Type]*seed.Snap, snapType snap.Type) (*snap.Info, error) {
+	seedSnap := sysSnaps[snapType]
+	mountPoint := filepath.Join(boot.InitramfsRunMntDir, snapTypeToMountDir[snapType])
+	hooksDir := filepath.Join(mountPoint, "meta", "hooks")
+	return snap.ReadInfoFromMountPoint(seedSnap.SnapName(), mountPoint, seedSnap.Path, hooksDir, seedSnap.SideInfo)
+}
+
+func runFDESetupHook(req *fde.SetupRequest) ([]byte, error) {
+	cmd := exec.Command("fde-setup")
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return make([]byte, 0), err
+	}
+	encoded, err := json.Marshal(req)
+	if err != nil {
+		return make([]byte, 0), err
+	}
+	writeErr := make(chan error)
+	go func() {
```

Comment:
I wonder if this could be simplified like this (go already supports a io.Reader for exec.Cmd and will essentially do the same internally as the existing code is doing):
```go
func runFDESetupHook(req *fde.SetupRequest) ([]byte, error) {
	encoded, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	cmd := exec.Command("fde-setup")
	cmd.Stdin = bytes.NewBuffer(encoded)
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}
	return output, nil
}
```
the diff:
```go
diff --git a/cmd/snap-bootstrap/cmd_initramfs_mounts.go b/cmd/snap-bootstrap/cmd_initramfs_mounts.go
index 6e0f6c2e67..0a6f81b767 100644
--- a/cmd/snap-bootstrap/cmd_initramfs_mounts.go
+++ b/cmd/snap-bootstrap/cmd_initramfs_mounts.go
@@ -20,6 +20,7 @@
 package main
 
 import (
+       "bytes"
        "crypto/subtle"
        "encoding/json"
        "errors"
@@ -255,31 +256,15 @@ func getSnap(sysSnaps map[snap.Type]*seed.Snap, snapType snap.Type) (*snap.Info,
 }
 
 func runFDESetupHook(req *fde.SetupRequest) ([]byte, error) {
-       cmd := exec.Command("fde-setup")
-       stdin, err := cmd.StdinPipe()
-       if err != nil {
-               return make([]byte, 0), err
-       }
        encoded, err := json.Marshal(req)
        if err != nil {
-               return make([]byte, 0), err
-       }
-       writeErr := make(chan error)
-       go func() {
-               var err error
-               func() {
-                       defer stdin.Close()
-                       _, err = stdin.Write(encoded)
-               }()
-               writeErr <- err
-       }()
-       output, err := cmd.Output()
-       if err != nil {
-               return make([]byte, 0), err
+               return nil, err
        }
-       err = <-writeErr
+       cmd := exec.Command("fde-setup")
+       cmd.Stdin = bytes.NewBuffer(encoded)
+       output, err := cmd.Output()
        if err != nil {
-               return make([]byte, 0), err
+               return nil, err
        }
        return output, nil
 }
 ```

Reply:
Yes, that looks better. I also wanted to know if we need to run this in `systemd-run`.

Reply:
yes, we should follow that plan and do this as some helper in kernel/fde as well